File: Book/3_primary_sort/ALDS1_2_D_Shell_Sort.cpp
#include <iostream>
#include <algorithm>
#include <stdio.h>
#include <vector>
using namespace std;
int cnt=0;
vector<int> G;

void insertionSort(int lst[], int n, int g){
    int i, j, v;
    for (i=g; i<n; i++){
        v = lst[i];
        j = i - g;
        while (j >= 0 && lst[j] > v){
            lst[j+g] = lst[j];
            j -= g;
            cnt++;
        }
        lst[j+g] = v;
    }
}

void shellSort(int lst[], int n){
    int h = 1, i;
    while (true){
        if (h>n) break;
        G.push_back(h);
        h = 3*h + 1;
    }
    for (i=G.size()-1; i>=0; i--){
        insertionSort(lst, n, G[i]);
    }
}

int main(){
    int i;
    int n, lst[1000000];
    cin >> n;
    for (i=0; i<n; i++) scanf("%d", &lst[i]);

    shellSort(lst, n);

    cout << G.size() << endl;
    for (i=G.size()-1; i>=0; i--){
        if (i == 0) printf("%d\n", G[i]);
        else printf("%d ", G[i]);
    }
    cout << cnt << endl;
    for (i=0; i<n; i++){
        printf("%d\n", lst[i]);
    }
}

File: AOJ/Book/3_primary_sort/ALDS1_2_D_Shell_Sort.py
#gだけ離れた要素のみを対象にした挿入ソート
def insertionSort(lst, n, g, cnt):
    for i in range(g,n):
        v = lst[i]
        j = i - g
        while j >= 0 and lst[j] > v:
            lst[j+g] = lst[j]
            j = j - g
            cnt += 1
        lst[j+g] = v
    return cnt

def gGenerator(n):
    G = [1]
    while True:
        g = 3 * G[-1] + 1
        if g > n:
            return G[::-1]
        G.append(g)

def shellSort(lst, n):
    cnt = 0
    #G(n+1) = G(n)+1
    G = gGenerator(n)
    print(len(G))
    print(*G)
    for i in range(len(G)):
        cnt = insertionSort(lst, n, G[i], cnt)
    print(cnt)
    return lst


n = int(input())
lst = []
for i in range(n):
    lst.append(int(input()))
lst_n = shellSort(lst, n)
for num in lst_n:
    print(num)

#挿入ソートの「ほぼ整列されたデータに対して高速に動作する」
#という長所を活かしたアルゴリズム


